rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+
+  Environments that are non-inductive could be created from the type as the
+  safe Rust type constraints describe a finite space.
+
+  For dealing with pointers one applicable mechanism could be *memory
+  predicates* to declaratively describe the state of the heap both before and
+  after the function call. 
+  
+  In CBMC's implementation memory predicates are part of the pre/postconditions.
+  This does not easily translate to Kani, since we handle pre/postconditions
+  manually and mainly in proc-macros. There are multiple ways to bridge this
+  gap, perhaps the easiest being to add memory predicates *separately* to Kani
+  instead of as part of pre/postconditions, so they can be handled by forwarding
+  them to CBMC. However this is also tricky, because memory predicates are used
+  to describe pointers and pointers only. Meaning that if they are encapsulated
+  in a structure (such as `Vec` or `RefCell`) there is no way of specifying the
+  target of the predicate without breaking encapsulation (similar to
+  `modifies`).
+  
+  A better solution would be for the data structure to declare its own
+  invariants at definition site which are automatically swapped in on every
+  contract that uses this type.
+- What about mutable trait inputs (wrt memory access patters), e.g. a `mut impl AccessMe`
+- **Trait contracts:** Our proposal could be extended easily to handle simple
+  trait contracts. The macros would generate new trait methods with default
+  implementations, similar to the functions it generates today. Using sealed
+  types we can prevent the user from overwriting the generated contract methods.
+  Contracts for the trait and contracts on it's `impl`s are combined by abstracting
+  the original method depending on context. The occurrence inside the contract
+  generated from the trait method is replaced by the `impl` contract. Any other
+  occurrence is replaced by the just altered trait method contract.
+- **Cross Session Verification Caching:** This proposal focuses on scalability
+  benefits within a single verification session, but those verification results
+  could be cached across sessions and speed up verification for large projects
+  using contacts in the future.
+- **Inductive Reasoning:** Describing recursive functions can require that the
+  contract also recurse, describing a fixpoint logic. This is needed for
+  instance for linked data structures like linked lists or trees.
+- **Compositional Contracts:** The proposal in this document lacks a
+  comprehensive handling of type parameters. Contract checking harnesses require
+  monomorphization. However this means the contract is only checked against a
+  finite number of instantiations of any type parameter (at most as many as
+  contract checking harnesses were defined). There is nothing preventing the
+  user from using different instantiations of the function's type parameters.
+
+  A function (`f()`) can only interact with its type parameters `P` through the
```

Comment:
Verus has an interesting way to encode polymrphism for contract checking, have you looked at their papers/code ?

Reply:
Not yet. I'll look into it if I get to the point where I have time to implement it.